pkg/scanner/scanner.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/weaveworks/profiles/pkg/gitrepository"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate -o fakes/fake_git_client.go . GitClient
+//GitClient client for interacting with git
+type GitClient interface {
+	ListTags(url string, secret *corev1.Secret) ([]string, error)
+}
+
+//counterfeiter:generate -o fakes/fake_repo_manager.go . GitRepositoryManager
+//GitRepositoryManager for managing gitrepositorys
+type GitRepositoryManager interface {
+	CreateAndWaitForResources(repo profilesv1.Repository, tags []gitrepository.Instance) ([]*sourcev1.GitRepository, error)
+	DeleteResources([]*sourcev1.GitRepository) error
+}
+
+//counterfeiter:generate -o fakes/fake_http_client.go . HTTPClient
+//HTTPClient for making HTTP requests
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+//Scanner for scanning repositorys
+type Scanner struct {
+	gitRepositoryManager GitRepositoryManager
+	gitClient            GitClient
+	httpClient           HTTPClient
+	logger               logr.Logger
+	namespace            string
+	ctx                  context.Context
+}
+
+//New returns a Scanner
+func New(gitRepositoryManager GitRepositoryManager, gitClient GitClient, httpClient HTTPClient, logger logr.Logger) *Scanner {
+	return &Scanner{
+		gitRepositoryManager: gitRepositoryManager,
+		gitClient:            gitClient,
+		httpClient:           httpClient,
+		logger:               logger,
+		namespace:            "profiles-system",
+		ctx:                  context.TODO(),
+	}
+}
+
+//ScanRepository for profiles
+func (s *Scanner) ScanRepository(repo profilesv1.Repository, secret *corev1.Secret) ([]profilesv1.ProfileCatalogEntry, error) {
+	tags, err := s.gitClient.ListTags(repo.URL, secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list tags: %w", err)
+	}
+	s.logger.Info("found tags", "url", repo.URL, "tags", tags)
+
+	var instances []gitrepository.Instance
+	for _, tag := range tags {
+		semver, path := getSemverAndPathFromTag(tag)
+		_, err := version.ParseVersion(semver)
+		if err == nil {
+			instances = append(instances, gitrepository.Instance{
+				Tag:  tag,
+				Path: path,
+			})
+		}
+	}
+
+	gitRepositoryResources, err := s.gitRepositoryManager.CreateAndWaitForResources(repo, instances)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create gitrepository resources: %w", err)
+	}
+	s.logger.Info("gitrepositorys created", "gitrepositories", gitRepositoryResources)
+
+	defer func() {
+		err := s.gitRepositoryManager.DeleteResources(gitRepositoryResources)
+		if err != nil {
+			s.logger.Error(err, "failed to cleanup git resources", "gitrepositories", gitRepositoryResources)
+		}
+	}()
+
+	var profiles []profilesv1.ProfileCatalogEntry
+	for _, gitRepo := range gitRepositoryResources {
+		profileDef, err := s.fetchProfileFromTarball(gitRepo)
+		if err != nil {
+			return nil, err
```

Comment:
fmt.Errorf here too please. :)

Reply:
this is an error from a local func, so its up to the local func to correctly format the error message (this is a pattern I tend to stick to). If it was calling an external package I would wrap it :smile:

Reply:
If the calling function properly handles and tells you what the problem was, then it's fine. Otherwise, you wouldn't know what failed and which git resource, right?

Reply:
Yeah! Here it passes in all the info it needs for it to correctly error. 

is only in a func for making the code cleaner, it could/should be able to get copied and pasted in line, and none of the error messages change :smile:

Reply:
Cool 👍